Lwm2m string api deprecation

jarlamsa opened the pull request:
### Description
LwM2M engine has used string based references for LwM2M paths in many places. Some API's use the `lwm2m_obj_path` -struct for paths. This PR deprecates the old API's and creates new functions which will take the `lwm2m_obj_path` as path parameter instead of a `const char*`.

The new API functions drop the `engine` part away from the function names. I.e. `lwm2m_engine_get_string` -> `lwm2m_get_string`. The old functions can still be used, but they will call the new API after the `const char*` path conversion to `lwm2m_obj_path` has been made.

### Usage
Old function usage:
`lwm2m_engine_create_res_inst("3/0/6/0");`
New function usage:
`lwm2m_engine_create_res_inst(&LWM2M_OBJ(3, 0, 6, 0));`

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/53209
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/53676

Commit:
net: lib: lwm2m: Deprecate string based set get API

Use the lwm2m_obj_path struct instead and deprecate old.

Signed-off-by: Jarno Lämsä <jarno.lamsa@nordicsemi.no>

Commit:
net: lib: lwm2m: Deprecate object and resource API

Deprecate old API and offer a new API for object and resource
creation and deletion. The new API uses path struct instead
of using a string as a reference to a path.

Signed-off-by: Jarno Lämsä <jarno.lamsa@nordicsemi.no>

Commit:
net: lib: lwm2m: Deprecate observation API

Add new APIs using the lwm2m path structs instead of
string references to paths.

Signed-off-by: Jarno Lämsä <jarno.lamsa@nordicsemi.no>

Commit:
net: lib: lwm2m: Deprecate callback registration API

Deprecate old API and create new API using path structs
instead of string references to paths.

Signed-off-by: Jarno Lämsä <jarno.lamsa@nordicsemi.no>

Commit:
net: lib: lwm2m: Deprecate buffer set get API

Deprecate old API and make new API using path structs
instead of using old string reference based paths.

Signed-off-by: Jarno Lämsä <jarno.lamsa@nordicsemi.no>

Commit:
net: lib: lwm2m: Replace deprecated function calls

Replace calls to deprecated functions with new ones.

Signed-off-by: Jarno Lämsä <jarno.lamsa@nordicsemi.no>

Commit:
samples: net: lwm2m_client: Replace deprecated calls

Replace calls to deprecated functions with new ones.

Signed-off-by: Jarno Lämsä <jarno.lamsa@nordicsemi.no>

Commit:
doc: Update LwM2M documentation

Update the documentation according to API changes.

Signed-off-by: Jarno Lämsä <jarno.lamsa@nordicsemi.no>

Commit:
net: lib: lwm2m: Deprecate string based send

Deprecate send API using the string references as paths.
Replace it with one using path structs.

Signed-off-by: Jarno Lämsä <jarno.lamsa@nordicsemi.no>

jarlamsa commented:
@juhaylinen @jheiskan81 @VeijoPesonen @carlescufi please review.

VeijoPesonen commented on include/zephyr/net/lwm2m.h:
```diff
@@ -1171,7 +1747,29 @@ int lwm2m_engine_get_res_buf(const char *pathstr, void **buffer_ptr, uint16_t *b
+ * If you directly write into the buffer, you must use lwm2m_engine_set_res_data_len()
+ * function to update the new size of the written data.
+ *
+ * All parameters except pathstr can NULL if you don't want to read those values.
```
```suggestion
 * All parameters, except for the pathstr, can be NULL if you don't want to read those values.
```

VeijoPesonen commented on include/zephyr/net/lwm2m.h:
```diff
@@ -1195,19 +1795,47 @@ int lwm2m_engine_get_res_data(const char *pathstr, void **data_ptr, uint16_t *da
+ *
+ * LwM2M clients use this function to create multi-resource instances:
+ * Example to create 0 instance of device available power sources:
+ * lwm2m_engine_create_res_inst(&LWM2M_OBJ(3, 0, 6, 0));
```
```suggestion
 * lwm2m_create_res_inst(&LWM2M_OBJ(3, 0, 6, 0));
```

VeijoPesonen reviewed (changes requested):
I have only checked lwm2m.h this far. I'm going to continue the review.

jarlamsa commented on subsys/net/lib/lwm2m/lwm2m_registry.c:
```diff
@@ -514,9 +545,8 @@ static int lwm2m_check_buf_sizes(uint8_t data_type, uint16_t resource_length, ui
 }
 
-static int lwm2m_engine_set(const char *pathstr, const void *value, uint16_t len)
+static int lwm2m_engine_set(const struct lwm2m_obj_path *path, const void *value, uint16_t len)
```
Well, it isn't public API similar to the `lwm2m_engine_get`.

jarlamsa commented on subsys/net/lib/lwm2m/lwm2m_registry.c:
```diff
@@ -526,37 +556,32 @@ static int lwm2m_engine_set(const char *pathstr, const void *value, uint16_t len
 		LOG_ERR("res instance data pointer is read-only "
-			"[%u/%u/%u/%u:%u]",
-			path.obj_id, path.obj_inst_id, path.res_id, path.res_inst_id, path.level);
+			"[%u/%u/%u/%u:%u]", path->obj_id, path->obj_inst_id, path->res_id,
```
Might be, my PR wasn't focused on the specifics on logging.

jarlamsa commented on subsys/net/lib/lwm2m/lwm2m_registry.c:
```diff
@@ -1357,6 +1826,19 @@ int lwm2m_engine_register_exec_callback(const char *pathstr, lwm2m_engine_execut
+	return lwm2m_register_exec_callback(&path, cb);
+}
+
 int lwm2m_engine_register_create_callback(uint16_t obj_id, lwm2m_engine_user_cb_t cb)
```
Good point, though these APIs aren't using pathstr, it would be good to harmonize the names.

VeijoPesonen commented on subsys/net/lib/lwm2m/lwm2m_registry.c:
```diff
@@ -847,45 +1048,63 @@ int lwm2m_engine_get_res_buf(const char *pathstr, void **buffer_ptr, uint16_t *b
 		LOG_ERR("path must have at least 3 parts");
 		return -EINVAL;
 	}
+	LOG_DBG("path:%u/%u/%u, buf:%p, buflen:%d", path->obj_id, path->obj_inst_id,
```
No actions are required from my side.

Commit:
net: lib: lwm2m: Deprecate string based enable cache

Deprecate the old API and replace with new one which uses
the lwm2m_obj_path struct instead of a string.

Signed-off-by: Jarno Lämsä <jarno.lamsa@nordicsemi.no>

jarlamsa commented:
Thank you @VeijoPesonen for your review. I think I have addressed all your comments. Please re-review.

jarlamsa commented:
Don't know what happened to the other reviewers as I requested a re-review from Veijo. @rlubos can you add them back as I can't add reviewers.

VeijoPesonen reviewed (dismissed):
Looks good to me.

SeppoTakalo commented on include/zephyr/net/lwm2m.h:
```diff
@@ -683,290 +753,658 @@ void lwm2m_registry_unlock(void);
+ *
+ * @return 0 for success or negative in case of error.
+ */
+int lwm2m_set_float(const struct lwm2m_obj_path *path, const double *value);
```
```suggestion
int lwm2m_set_f64(const struct lwm2m_obj_path *path, const double value);
```

jarlamsa replied:
Done

SeppoTakalo commented on include/zephyr/net/lwm2m.h:
```diff
@@ -683,290 +753,658 @@ void lwm2m_registry_unlock(void);
  * @return 0 for success or negative in case of error.
  */
-int lwm2m_engine_get_float(const char *pathstr, double *buf);
+int lwm2m_get_float(const struct lwm2m_obj_path *path, double *buf);
```
```suggestion
int lwm2m_get_f64(const struct lwm2m_obj_path *path, double *value);
```

jarlamsa replied:
Done

SeppoTakalo reviewed (changes requested):
Because we are touching this API now, I  would like to see that we change one minor thing from floating point getter/setter.
* Set should use value in parameter, not pointer to value. Same as others.
* `set_f64` instead of `set_float()` that takes double as a parameter. All other use s/uBITS naming. This could too.
* In getter, point to `*value` not `*buffer`, this is a single intem.

jarlamsa commented:
@SeppoTakalo I have addressed your review comments, please re-review.

rlubos commented on doc/connectivity/networking/api/lwm2m.rst:
```diff
@@ -413,16 +413,16 @@ For a more detailed LwM2M client sample see: :ref:`lwm2m-client-sample`.
-  lwm2m_engine_set_float("3303/0/5700", &value);
+  lwm2m_set_u32(&LWM2M_OBJ(1, 0, 1), 60);
+  lwm2m_set_u8(&LWM2M_OBJ(5, 0, 3), 0);
+  lwm2m_set_float(&LWM2M_OBJ(3303, 0, 5700), &value);
```
This should be `lwm2m_set_f64` now, right?

jarlamsa replied:
Done

rlubos reviewed (commented):
Looks good, just one nit spotted. It would be good if you could mention that LwM2M APIs has been deprecated [here](https://github.com/zephyrproject-rtos/zephyr/blame/main/doc/releases/release-notes-3.3.rst#L110).

VeijoPesonen commented:
There are still typedefs etc. which have the prefix _lwm2m_engine_. My assumption is that those are handled in a separate PR, or it might be there's no need to touch those.

rlubos reviewed (dismissed):
Thanks!

jarlamsa commented:
Rebased to retrigger failing CI
